### PR TITLE
Migrate from emscripten to wasi-sdk for the wasm development tool chain

### PR DESCRIPTION
**Pull request by cmickeyb**
Deprecate the use of emscripten for developing contracts for wawaka. We have struggled to support current version of emscripten. In an attempt to be more consistent with WAMR, this PR moves the toolchain to [WASI-SDK](https://github.com/WebAssembly/wasi-sdk). 

The migration enables the use of many classes from the C++ std library (strings, vectors, ...) which should make it significantly easier to write wawaka contracts.

### CODE REVIEW

**Review by bvavala — APPROVED** [review]
lgtm -- tested in sim-mode in a non-dockerized env.

**Review by bvavala — CHANGES_REQUESTED** [review]
I run the tests  in a non-dockerized environment and there are some issues with the build because, in my case, WASI sdk is not the expected path. In particular,
* WAMR complains that it cannot find it
* cmake tries to run some compiler tests but does not find some libs/objs
* our contract build fails as it cannot find some includes

This commit https://github.com/bvavala/private-data-objects/commit/8a09cb8d6f31e1650658cf2e051eebe9119b7f3c fixes the issues (at least on my machine) and CI passes too.

File: contracts/wawaka/contract-build.cmake [hunk 1]
```diff
@@ -15,75 +15,126 @@
...
+  MESSAGE(FATAL_ERROR "WASM_MEM_CONFIG environment variable not defined!")
+ENDIF()
+SET(WASM_SRC "$ENV{WASM_SRC}")
 
 IF (NOT DEFINED ENV{WASM_MEM_CONFIG})
   MESSAGE(FATAL_ERROR "WASM_MEM_CONFIG environment variable not defined!")
 ENDIF()
+SET(WASM_MEM_CONFIG "$ENV{WASM_MEM_CONFIG}")
+
+IF (NOT DEFINED ENV{WASI_SDK_DIR})
```
Comment (masomel) [inline contracts/wawaka/contract-build.cmake:36]: We should probably document this variable to docs/environment.md, no?

**Review by masomel — APPROVED** [review]
LGTM. Two small comments around the env variables.

File: contracts/wawaka/contract-build.cmake (hunk 1 above)
Comment (cmickeyb) [reply contracts/wawaka/contract-build.cmake]: removed the requirement for the environment variable.

File: contracts/wawaka/contract-build.cmake [hunk 2]
```diff
@@ -15,75 +15,126 @@
...
+SET(PDO_HOME "$ENV{PDO_HOME}")
+
+IF(NOT EXISTS $ENV{PDO_SOURCE_ROOT})
+  MESSAGE(FATAL "PDO_SOURCE_ROOT environment variable not defined")
+ENDIF()
+SET(PDO_SOURCE_ROOT $ENV{PDO_SOURCE_ROOT})
+SET(WAWAKA_ROOT ${PDO_SOURCE_ROOT}/contracts/wawaka)
+
+IF (NOT DEFINED ENV{WASM_SRC})
+  MESSAGE(FATAL_ERROR "WASM_MEM_CONFIG environment variable not defined!")
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply contracts/wawaka/contract-build.cmake]: fixed.

File: common/interpreter/wawaka_wasm/README.md [hunk 3]
```diff
@@ -16,81 +16,69 @@ set of supporting functions that can be executed inside the contract
...
-
-source ./emsdk_env.sh
+wget -q -P /tmp https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-12/wasi-sdk_12.0_amd64.deb
+sudo dpkg --install /tmp/wasi-sdk_12.0_amd64.deb
 ```
 
-### WAMR setup ###
+These directions assume that the SDK will be installed in the PDO source tree. Typically, the WASI
+SDK would be installed in the directory `/opt/wasi-sdk`. Please set the environment variable
+`WASI_SDK_DIR` to the directory where the SDK is installed.
```
Comment (bvavala) [inline common/interpreter/wawaka_wasm/README.md:43]: Although this works, the `WASI_SDK_DIR` environment variable set by the user is ignored.
In fact, the default value of the `WASI_SDK_DIR` _cmake_ variable (i.e., `/opt/wasi-sdk`) is always set by the WAMR toolchain.

So, again,
* either we set the cmake variable in the Makefile (and the WAMR toolchain will use it),
* or (suggestion below) we remove the environment variable as it is not required. So we will be relying on the fact that the debian package installation folder and the default value in the WAMR toolchain match.

```suggestion
These directions assume that the WASI SDK will be installed in the default directory `/opt/wasi-sdk`.
```

Comment (cmickeyb) [reply common/interpreter/wawaka_wasm/README.md:43]: you should never set wasi_sdk_dir. it is not used at all.

**Review by bvavala — APPROVED** [review]
It works and the instructions are good.

For the record, the `WASI_SDK_DIR` cmake variable is set by the WAMR toolchain to `/opt/wasi-sdk`. So we rely on the fact that the debian package installation folder _is_ `/opt/wasi-sdk`.

**Review by prakashngit — APPROVED** [review]

**Pull request merged**